### ISSUE TITLE
Language server raises exception when client does not have any configuration settings to provide it.

### ISSUE THREAD

**Issue by cowboyd** [microsoft/vscode-eslint#2002]
Error: Request textDocument/codeAction failed with message: Cannot read properties of null (reading 'workingDirectory')

I'm [trying to get the `vscode-eslint` server working](https://github.com/thefrontside/lspx/issues/3) with the Emacs `eglot` client. However, it is raising an exception because of what is apparently a `null` settings for the requested file. If I understand [the spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_configuration) correctly however, Eglot's  behavior is ok.

>  If the client can’t provide a configuration setting for a given scope then null needs to be present in the returned array.

Here is the full transcript of the conversation between Eglot and `vscode-eslint`. As you can see at the end, it apparently requests the configuration as part of responding to the `textDocument/codeAction` request. When Eglot responds that is has no further configuration available by returning `null`, the `textDocument/codeAction` request fails with the above exception.

One further note: Eglot does issue this warning as part of the transcript: `[eglot] (warning) Server tried to register unsupported capability ‘workspace/didChangeWorkspaceFolders’`.

```

 --> [client2server.request] [ "vscode-eslint-language-server" ] initialize [
   {
     processId: 34058,
     clientInfo: { name: "Eglot", version: "1.18.0.20250408.74617" },
     rootPath: "/Users/cowboyd/Code/@cowboyd/react-ts-eslint-prettier-sample/",
     rootUri: "file:///Users/cowboyd/Code/@cowboyd/react-ts-eslint-prettier-sample",
     initializationOptions: {},
     capabilities: {
       workspace: {
         applyEdit: true,
         executeCommand: { dynamicRegistration: false },
         workspaceEdit: { documentChanges: true },
         didChangeWatchedFiles: { dynamicRegistration: true },
         symbol: { dynamicRegistration: false },
         configuration: true,
         workspaceFolders: true
       },
       textDocument: {
         synchronization: {
           dynamicRegistration: false,
           willSave: true,
           willSaveWaitUntil: true,
           didSave: true
         },
         completion: {
           dynamicRegistration: false,
           completionItem: [Object],
           contextSupport: true
         },
         hover: { dynamicRegistration: false, contentFormat: [Array] },
         signatureHelp: { dynamicRegistration: false, signatureInformation: [Object] },
         references: { dynamicRegistration: false },
         definition: { dynamicRegistration: false, linkSupport: true },
         declaration: { dynamicRegistration: false, linkSupport: true },
         implementation: { dynamicRegistration: false, linkSupport: true },
         typeDefinition: { dynamicRegistration: false, linkSupport: true },
         documentSymbol: {
           dynamicRegistration: false,
           hierarchicalDocumentSymbolSupport: true,
           symbolKind: [Object]
         },
         documentHighlight: { dynamicRegistration: false },
         codeAction: {
           dynamicRegistration: false,
           resolveSupport: [Object],
           dataSupport: true,
           codeActionLiteralSupport: [Object],
           isPreferredSupport: true
         },
         formatting: { dynamicRegistration: false },
         rangeFormatting: { dynamicRegistration: false },
         rename: { dynamicRegistration: false },
         inlayHint: { dynamicRegistration: false },
         callHierarchy: { dynamicRegistration: false },
         typeHierarchy: { dynamicRegistration: false },
         publishDiagnostics: {
           relatedInformation: false,
           codeDescriptionSupport: false,
           tagSupport: [Object]
         }
       },
       window: {
         showDocument: { support: true },
         showMessage: { messageActionItem: [Object] },
         workDoneProgress: true
       },
       general: { positionEncodings: [ "utf-32", "utf-8", "utf-16" ] },
       experimental: {}
     },
     workspaceFolders: [
       {
         uri: "file:///Users/cowboyd/Code/@cowboyd/react-ts-eslint-prettier-sample",
         name: "~/Code/@cowboyd/react-ts-eslint-prettier-sample/"
       }
     ]
   },
   MutableToken { _isCancelled: false }
 ]
 --> [server2client.notify] vscode-eslint-language-server window/logMessage [ { type: 3, message: "ESLint server running in node v22.9.0" } ]
 <-- [client2server.request] [ "vscode-eslint-language-server" ] initialize {
   serverInfo: { name: "lspx", version: "0.1.0" },
   capabilities: {
     textDocumentSync: {
       openClose: true,
       change: 2,
       willSaveWaitUntil: false,
       save: { includeText: false }
     },
     workspace: { workspaceFolders: { supported: true } },
     executeCommandProvider: {
       commands: [
         "eslint.applySingleFix",
         "eslint.applySuggestion",
         "eslint.applySameFixes",
         "eslint.applyAllFixes",
         "eslint.applyDisableLine",
         "eslint.applyDisableFile",
         "eslint.openRuleDoc"
       ]
     },
     diagnosticProvider: {
       identifier: "eslint",
       interFileDependencies: false,
       workspaceDiagnostics: false
     },
     codeActionProvider: { codeActionKinds: [ "quickfix", "source.fixAll.eslint" ] }
   }
 }
 --> [client2server.notify] [ "vscode-eslint-language-server" ] initialized [ {} ]
 --> [client2server.notify] [ "vscode-eslint-language-server" ] textDocument/didOpen [
   {
     textDocument: {
       uri: "file:///Users/cowboyd/Code/@cowboyd/react-ts-eslint-prettier-sample/src/App.tsx",
       version: 0,
       languageId: "typescript",
       text: 'import { useState } from "react";\n' +
         'import reactLogo from "./assets/react.svg";\n' +
         'import viteLogo from "/vite.svg";\n' +
         'import "./App.css";\n' +
         "\n" +
         "const emptyFunc = (_, a) => {\n" +
         "  return a;\n" +
         "};\n" +
         "\n" +
         "function App() {\n" +
         "           const [count, setCount] = useState(0);\n" +
         "\n" +
         "       const x: number = '1234';\n" +
         "\n" +
         "  return (\n" +
         "          <>\n" +
         "      <div>\n" +
         '        <a href="https://vite.dev" target="_blank">\n' +
         '          <img src={viteLogo} className="logo" alt="" />\n' +
         "        </a>\n" +
         '        <a href="https://react.dev" target="_blank">\n' +
         '          <img src={reactLogo} className="logo react" alt="React logo" />\n' +
         "        </a>\n" +
         "      </div>\n" +
         "      <h1>Vite + React</h1>\n" +
         '      <div className="card">\n' +
         "        <button onClick={() => setCount((count) => count + 1)}>\n" +
         "          count is {count}\n" +
         "        </button>\n" +
         "        <p>\n" +
         "          Edit <code>src/App.tsx</code> and save to test HMR\n" +
         "        </p>\n" +
         "      </div>\n" +
         '      <p className="read-the-docs">\n' +
         "        Click on the Vite and React logos to learn more\n" +
         "      </p>\n" +
         "    </>\n" +
         "  );\n" +
         "}\n" +
         "\n" +
         "export default App;\n"
     }
   }
 ]
 --> [client2server.notify] [ "vscode-eslint-language-server" ] workspace/didChangeConfiguration [ { settings: {} } ]
 --> [server2client.request] vscode-eslint-language-server client/registerCapability [
   {
     registrations: [
       {
         id: "a9cd965b-6aff-4307-930d-ba22fb57dde0",
         method: "workspace/didChangeWorkspaceFolders",
         registerOptions: {}
       }
     ]
   },
   MutableToken { _isCancelled: false }
 ]
 <-- [server2client.request] vscode-eslint-language-server client/registerCapability null
 --> [client2server.request] [ "vscode-eslint-language-server" ] textDocument/codeAction [
   {
     textDocument: {
       uri: "file:///Users/cowboyd/Code/@cowboyd/react-ts-eslint-prettier-sample/src/App.tsx"
     },
     range: {
       start: { line: 34, character: 51 },
       end: { line: 34, character: 55 }
     },
     context: { diagnostics: [], triggerKind: 2 }
   },
   MutableToken { _isCancelled: false }
 ]
 --> [server2client.request] vscode-eslint-language-server workspace/configuration [
   {
     items: [
       {
         scopeUri: "file:///Users/cowboyd/Code/@cowboyd/react-ts-eslint-prettier-sample/src/App.tsx",
         section: ""
       }
     ]
   },
   MutableToken { _isCancelled: false }
 ]
 <-- [server2client.request] vscode-eslint-language-server workspace/configuration [ null ]
 Error: Request textDocument/codeAction failed with message: Cannot read properties of null (reading 'workingDirectory')
     at handleResponse (file:///var/folders/zs/m38wmycs4pj6ymtkk3cjrp0w0000gn/T/deno-compile-lspx/.deno_compile_node_modules/localhost/vscode-jsonrpc/8.2.1/lib/common/connection.js:565:48)
     at handleMessage (file:///var/folders/zs/m38wmycs4pj6ymtkk3cjrp0w0000gn/T/deno-compile-lspx/.deno_compile_node_modules/localhost/vscode-jsonrpc/8.2.1/lib/common/connection.js:345:13)
     at processMessageQueue (file:///var/folders/zs/m38wmycs4pj6ymtkk3cjrp0w0000gn/T/deno-compile-lspx/.deno_compile_node_modules/localhost/vscode-jsonrpc/8.2.1/lib/common/connection.js:362:17)
     at file:///var/folders/zs/m38wmycs4pj6ymtkk3cjrp0w0000gn/T/deno-compile-lspx/.deno_compile_node_modules/localhost/vscode-jsonrpc/8.2.1/lib/common/connection.js:334:13
     at ext:deno_web/02_timers.js:35:14
     at eventLoopTick (ext:core/01_core.js:213:13) {
   code: -32603,
   data: undefined
 }
```

**Comment by cowboyd** [comment]
@dbaeumer I'd be happy to make a PR if you can point me to where the problem is happening (I assume that it's some missing configuration), and what you think the defaults I need to provide are.

**Comment by dbaeumer** [comment]
The problem is that the server expects a working directory setting when receiving the settings from the client. So the server needs to be changed to handle the case that it is `null` or not present. The general code with dealing with setting is here: https://github.com/microsoft/vscode-eslint/blob/c0e753713ea9935667e849d91e549adbff213e7e/server/src/eslint.ts#L851

**Comment by cowboyd** [comment]
@dbaeumer Nice catch! I totally missed the `../`. Getting rid of that rogue `node_modules/` directory worked.

**Comment by SCWells72** [comment]
Did you ever get past this error? I'm trying to use this language server with another client (LSP4IJ), and I'm seeing the exact same error. I don't know if I need to raise an issue with LSP4IJ to ensure it sends over a working directory somehow, or if there's a configuration change required, or what. Interested to know if you've been able to get this working with eglot (or any other non-VS Code client).

**Comment by cowboyd** [comment]
@SCWells72 I wasn't able to get close by looking at how the nvim language server handled it, but I never got it quite working

**Comment by angelozerr** [comment]
@dbaeumer @cowboyd @SCWells72 for your information I have integrated (basicly) eslint language server inside IntelliJ with LSP4IJ

See https://github.com/redhat-developer/lsp4ij/blob/main/docs/user-defined-ls/vscode-eslint-language-server.md

To avoid the original error, the fix was to send the following configuration to the eslint language server:

```json
{
  "validate": "on",
  "packageManager": "npm",
  "useESLintClass": false,
  "experimental": {
    "useFlatConfig": false
  },
  "codeActionOnSave": {
    "mode": "all"
  },
  "format": false,
  "quiet": false,
  "onIgnoredFiles": "off",
  "options": {},
  "rulesCustomizations": [],
  "run": "onType",
  "problems": {
    "shortenToSingleLine": false
  },
  "nodePath": "",
  "codeAction": {
    "disableRuleComment": {
      "enable": true,
      "location": "separateLine",
      "commentStyle": "line"
    },
    "showDocumentation": {
      "enable": true
    }
  }
}
```

The important config is   `"nodePath": "",` and not   "nodePath": null, which is the default value.

This integration of eslint was done with zero code, just with declarative mean, see JSON descriptors at https://github.com/redhat-developer/lsp4ij/tree/main/src/main/resources/templates/lsp/vscode-eslint-language-server

This config auto-install eslint language server with this installer https://github.com/redhat-developer/lsp4ij/blob/05c06d3e156530e2c45d58aacb3fd03358f513df/src/main/resources/templates/lsp/vscode-eslint-language-server/installer.json#L21 

`npm install vscode-langservers-extracted --force`

The integration of eslint language server inside IntelliJ doesn't provide the full features like vscode-eslint because the language server is not full compliant compatible with LSP:

 * client register custom file watcher https://github.com/microsoft/vscode-eslint/blob/c7c2ebf24c9e09313ab12a0792887d83b4e82827/client/src/client.ts#L426  those file watcher should be declared on server side.
 * client manages pull diagnostics of focus https://github.com/microsoft/vscode-eslint/blob/c7c2ebf24c9e09313ab12a0792887d83b4e82827/client/src/client.ts#L451, LSP doesn't provides this capability. Perhaps LSP should specifiy textDocument/didFocus because I have see that for language server wich manages AI.
 * other?

@dbaeumer it would be more cleaner to provide an npm module for eslint (instead of using the  vscode-langservers-extracted) and manage most features with LSP server.

Do you think you could add in your README, LSP client which integrates your LSP eslint language server? If you are OK I could create a PR to add LSP4IJ.

**Comment by SCWells72** [comment]
@angelozerr, the ESLint integration seems to working very well to me. I've made a few changes in my own plugin to accommodate it, specifically to disable another, slower ESLint integration (via Salesforce Code Analyzer 5) if that language server is installed/started. Combined with the TypeScript language server, this provides a really nice, complete set of static code analysis capabilities for JavaScript and TypeScript in Community Edition IDEs. Thanks so much!

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr), the ESLint integration seems to working very well to me. I've made a few changes in my own plugin to accommodate it, specifically to disable another, slower ESLint integration (via Salesforce Code Analyzer 5) if that language server is installed/started. Combined with the TypeScript language server, this provides a really nice, complete set of static code analysis capabilities for JavaScript and TypeScript in Community Edition IDEs. Thanks so much!

You are welcome @SCWells72 !

The last thing to do is to write (generate) the JSON Schema for the configuration. I wonder why ESLint langugae server doesn't use eslint as root JSON object settings?

**Comment by SCWells72** [comment]
> The last thing to do is to write (generate) the JSON Schema for the configuration. I wonder why ESLint langugae server doesn't use eslint as root JSON object settings?

I wondered that as well while tinkering with the language server config and consulting the language server docs. It doesn't seem to be the same flattened version of the same info without that root level. Either way, it's great to have this available in Community Edition IDEs in near-real-time now, though! Great work!

**Comment by n1ckisthereu** [comment]
Hello everyone I saw that this item has been open since last year, and I've been researching a bit to see the progress. I'd love to make eslint-lsp more compatible outside of VS Code, and I saw that you're open to a PR. Today I invested a few hours, but I didn't make any progress. I'm trying to get lsp to work with eglot in Emacs, like @cowboyd . If anyone could help me better understand the problem and provide a detailed solution, I'd be happy to continue this discussion.
